### ISSUE TITLE
terraform init --backend-config cannot update backend configuration

### ISSUE THREAD

**Issue by woz5999** [hashicorp/terraform#12834]
### Terraform Version
Terraform v0.9.0

### Affected Resource(s)
s3 backend partial config

Note that I haven't tested with other backends, so it may not be exclusive to this backend.

### Terraform Configuration Files
```hcl
terraform {
  backend "s3" {
    bucket = "my-bucket"
    region = "us-west-1"
  }
}
```

### Expected Behavior
What should have happened?

terraform init should update the backend configuration if the contents of the specified backend-config file changes. 

### Actual Behavior
What actually happened?
Once the backend configuration is initialized, it doesn't respect partial configuration updates

### Steps to Reproduce
Please list the steps required to reproduce the issue, for example:
```
echo "key = \"key1\"" > ./key.conf
terraform init --backend-config=./key.conf
grep "key" .terraform/terraform.tfstate
# "key": "key1"

echo "key = \"key2\"" > ./key.conf
terraform init --backend-config=./key.conf
grep "key" .terraform/terraform.tfstate
# "key": "key1"
```

**Comment by mitchellh** [comment]
This is fixed in master, will be part of 0.9.1 released today!

**Comment by woz5999** [comment]
Brilliant! Thanks.
